source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -0,0 +1,260 @@
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <deque>
+#include <functional>
+#include <string>
+#include <vector>
+
+#include "envoy/common/pure.h"
+#include "envoy/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.pb.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/registry/registry.h"
+#include "envoy/ssl/context.h"
+#include "envoy/ssl/context_config.h"
+#include "envoy/ssl/private_key/private_key.h"
+#include "envoy/ssl/ssl_socket_extended_info.h"
+
+#include "common/common/matchers.h"
+#include "common/common/regex.h"
+#include "common/config/datasource.h"
+#include "common/config/utility.h"
+#include "common/protobuf/message_validator_impl.h"
+#include "common/stats/symbol_table_impl.h"
+
+#include "extensions/transport_sockets/tls/cert_validator/factory.h"
+#include "extensions/transport_sockets/tls/cert_validator/well_known_names.h"
+#include "extensions/transport_sockets/tls/stats.h"
+#include "extensions/transport_sockets/tls/utility.h"
+
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+
+using SPIFFEConfig = envoy::extensions::transport_sockets::tls::v3::SPIFFECertValidatorConfig;
+
+SPIFFEValidator::SPIFFEValidator(const Envoy::Ssl::CertificateValidationContextConfig* config,
+                                 SslStats& stats, TimeSource& time_source)
+    : stats_(stats), time_source_(time_source) {
+  ASSERT(config != nullptr);
+
+  SPIFFEConfig message;
+  Config::Utility::translateOpaqueConfig(config->customValidatorConfig().value().typed_config(),
+                                         ProtobufWkt::Struct(),
+                                         ProtobufMessage::getStrictValidationVisitor(), message);
+
+  auto size = message.trust_domains().size();
+  trust_bundle_stores_.reserve(size);
+  for (auto& domain : message.trust_domains()) {
+    if (trust_bundle_stores_.find(domain.name()) != trust_bundle_stores_.end()) {
+      throw EnvoyException(absl::StrCat(
+          "Multiple trust bundles are given for one trust domain for ", domain.name()));
+    }
+
+    auto cert = Config::DataSource::read(domain.trust_bundle(), true, config->api());
+    bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(cert.data()), cert.size()));
+    RELEASE_ASSERT(bio != nullptr, "");
+    bssl::UniquePtr<STACK_OF(X509_INFO)> list(
+        PEM_X509_INFO_read_bio(bio.get(), nullptr, nullptr, nullptr));
+    if (list == nullptr || sk_X509_INFO_num(list.get()) == 0) {
+      throw EnvoyException(
+          absl::StrCat("Failed to load trusted CA certificate for ", domain.name()));
+    }
+
+    auto store = X509StorePtr(X509_STORE_new());
+    bool has_crl = false;
+    bool ca_loaded = false;
+    for (const X509_INFO* item : list.get()) {
+      if (item->x509) {
+        X509_STORE_add_cert(store.get(), item->x509);
+        X509_up_ref(item->x509);
```

Comment:
nit: I assume this added ref count is for storing in `ca_certs_`? I'm not familiar with Envoy policies on exception handling due to out-of-memory situations, but if the ref is added after the call to push_back() then `item-x509` would still be properly freed if an exception were to occur with the `push_back()`.

Reply:
we've already reserve the memory for map before entering this for-loop, so no relocation happens  in push_back which is exception-free. So I think this is fine.

Reply:
resolved in  https://github.com/envoyproxy/envoy/pull/14884/commits/cdfb3ced2b5125432a40c56697164eedab8f3d15